flink-state-backends/flink-statebackend-forst/src/test/java/org/apache/flink/state/forst/ForStAsyncAndSyncCompatibilityTest.java

```diff
@@ -0,0 +1,234 @@
+                    true);
+        } finally {
+            IOUtils.closeQuietly(syncKeyedStateBackend);
+        }
+    }
+
+    @Test
+    void testForStTransFromSyncToAsync() throws Exception {
+        ForStSyncKeyedStateBackend<String> keyedBackend =
+                createSyncKeyedStateBackend(
+                        forStStateBackend, env, StringSerializer.INSTANCE, Collections.emptyList());
+        org.apache.flink.api.common.state.MapStateDescriptor<Integer, String> descriptor =
+                new org.apache.flink.api.common.state.MapStateDescriptor<>(
+                        "testState", IntSerializer.INSTANCE, StringSerializer.INSTANCE);
+        org.apache.flink.api.common.state.MapState<Integer, String> mapState =
+                keyedBackend.getOrCreateKeyedState(IntSerializer.INSTANCE, descriptor);
+        keyedBackend.setCurrentKey("testKey");
+        ((InternalKvState) mapState).setCurrentNamespace(1);
+        mapState.put(1, "1");
+
+        RunnableFuture<SnapshotResult<KeyedStateHandle>> snapshot =
+                keyedBackend.snapshot(
+                        1L,
+                        System.currentTimeMillis(),
+                        env.getCheckpointStorageAccess()
+                                .resolveCheckpointStorageLocation(
+                                        1L, CheckpointStorageLocationReference.getDefault()),
+                        CheckpointOptions.forCheckpointWithDefaultLocation());
+
+        if (!snapshot.isDone()) {
+            snapshot.run();
+        }
+        SnapshotResult<KeyedStateHandle> snapshotResult = snapshot.get();
+        KeyedStateHandle stateHandle = snapshotResult.getJobManagerOwnedSnapshot();
+        IOUtils.closeQuietly(keyedBackend);
+
+        ForStKeyedStateBackend<String> asyncKeyedStateBackend =
+                setUpAsyncKeyedStateBackend(Collections.singletonList(stateHandle));
+
+        MapStateDescriptor<Integer, String> newStateDescriptor =
+                new MapStateDescriptor<>(
+                        "testState", IntSerializer.INSTANCE, StringSerializer.INSTANCE);
+        try {
+            MapState<Integer, String> asyncMapState =
+                    asyncKeyedStateBackend.createState(
+                            1, IntSerializer.INSTANCE, newStateDescriptor);
+            fail();
+
+            context = aec.buildContext("testRecord", "testKey");
+            context.retain();
+            aec.setCurrentContext(context);
+            asyncMapState
+                    .asyncGet(1)
+                    .thenAccept(
+                            value -> {
+                                Assert.assertEquals(value, "1");
+                            });
+            context.release();
+            aec.drainInflightRecords(0);
+        } catch (Exception e) {
```

Comment:
could we have a more specific exception and /or message. It is not obvious from the exception text that this means,  ForStStateBackend does not support switching from Sync to Async, so this.